core/web/chains_controller.go

```diff
@@ -0,0 +1,182 @@
+package web
+
+import (
+	"context"
+	"database/sql"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+	"github.com/manyminds/api2go/jsonapi"
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/chains"
+)
+
+type ChainSet[I chains.ID, C chains.Config] interface {
+	Add(ctx context.Context, id I, cfg C) (chains.Chain[I, C], error)
+	Show(id I) (chains.Chain[I, C], error)
+	Configure(ctx context.Context, id I, enabled bool, cfg C) (chains.Chain[I, C], error)
+	Remove(id I) error
+	Index(offset, limit int) ([]chains.Chain[I, C], int, error)
+}
+
+type ChainsController interface {
+	// Index lists chains.
+	Index(c *gin.Context, size, page, offset int)
+	// Create adds a new chain from a CreateChainRequest.
+	Create(*gin.Context)
+	// Show gets a chain by id.
+	Show(*gin.Context)
+	// Update configures an existing chain from an UpdateChainRequest.
+	Update(*gin.Context)
+	// Delete removes a chain.
+	Delete(*gin.Context)
+}
+
+type chainsController[I chains.ID, C chains.Config, R jsonapi.EntityNamer] struct {
+	resourceName  string
+	chainSet      ChainSet[I, C]
+	errNotEnabled error
+	parseChainID  func(string) (I, error)
+	newResource   func(chains.Chain[I, C]) R
+}
+
+func NewChainsController[I chains.ID, C chains.Config, R jsonapi.EntityNamer](prefix string, chainSet ChainSet[I, C], errNotEnabled error,
+	parseChainID func(string) (I, error), newResource func(chains.Chain[I, C]) R) *chainsController[I, C, R] {
+	return &chainsController[I, C, R]{
+		resourceName:  prefix + "_chain",
+		chainSet:      chainSet,
+		errNotEnabled: errNotEnabled,
+		parseChainID:  parseChainID,
+		newResource:   newResource,
+	}
+}
+
+func (cc *chainsController[I, C, R]) Index(c *gin.Context, size, page, offset int) {
+	if cc.chainSet == nil {
+		jsonAPIError(c, http.StatusBadRequest, cc.errNotEnabled)
+		return
+	}
+	chains, count, err := cc.chainSet.Index(offset, size)
+
+	if err != nil {
+		jsonAPIError(c, http.StatusBadRequest, err)
+		return
+	}
+
+	var resources []R
+	for _, chain := range chains {
+		resources = append(resources, cc.newResource(chain))
+	}
+
+	paginatedResponse(c, cc.resourceName, size, page, resources, count, err)
+}
+
+type CreateChainRequest[I chains.ID, C chains.Config] struct {
+	ID     I `json:"chainID"`
+	Config C `json:"config"`
+}
+
+func NewCreateChainRequest[I chains.ID, C chains.Config](id I, config C) CreateChainRequest[I, C] {
+	return CreateChainRequest[I, C]{ID: id, Config: config}
+}
+
+func (cc *chainsController[I, C, R]) Create(c *gin.Context) {
+	if cc.chainSet == nil {
+		jsonAPIError(c, http.StatusBadRequest, cc.errNotEnabled)
+		return
+	}
+
+	var request CreateChainRequest[I, C]
+	//TODo does this already handle utils.Big....
```

Comment:
A leftover or incomplete?

Reply:
:facepalm: Leftover, since the tests are passing. Will remove.